System/Random.hs

```diff
@@ -386,17 +402,84 @@ splitPrimGen = atomicPrimGen split >=> restore
 
 runPrimGenST :: RandomGen g => g -> (forall s . PrimGen s g -> ST s a) -> (a, g)
 runPrimGenST g action = runST $ do
-  primGen :: PrimGen s g <- restore g
+  primGen <- restore g
   res <- action primGen
   g' <- save primGen
   pure (res, g')
 
+-- | Same as `runPrimGenST`, but discard the resulting generator.
+runPrimGenST_ :: RandomGen g => g -> (forall s . PrimGen s g -> ST s a) -> a
+runPrimGenST_ g action = fst $ runPrimGenST g action
+
 runPrimGenIO :: (RandomGen g, MonadIO m) => g -> (PrimGen RealWorld g -> m a) -> m (a, g)
 runPrimGenIO g action = do
-  primGen :: PrimGen s g <- liftIO $ restore g
+  primGen <- liftIO $ restore g
   res <- action primGen
   g' <- liftIO $ save primGen
   pure (res, g')
+{-# INLINE runPrimGenIO #-}
+
+-- | Same as `runPrimGenIO`, but discard the resulting generator.
+runPrimGenIO_ :: (RandomGen g, MonadIO m) => g -> (PrimGen RealWorld g -> m a) -> m a
+runPrimGenIO_ g action = fst <$> runPrimGenIO g action
+{-# INLINE runPrimGenIO_ #-}
+
+
+newtype MutGen s g = MutGen (MutableByteArray s)
+
+instance (s ~ PrimState m, PrimMonad m, RandomGen g, Prim g) =>
```

Comment:
How difficult is it in practice to implement `Prim` for e.g. a type isomorphic to `(Word64, Word64)` (as is the case for `SMGen`)?

Reply:
`Storable` seems much easier:

```
instance Storable SMGen where
  sizeOf (SMGen seed gamma) = sizeOf seed + sizeOf gamma
  alignment (SMGen seed _gamma) = alignment seed
  peek ptr = do
    let ptr' = castPtr ptr
    seed  <- peekElemOff ptr' 0
    gamma <- peekElemOff ptr' 1
    return $ SMGen seed gamma
  poke ptr (SMGen seed gamma) = do
    let ptr' = castPtr ptr
    pokeElemOff ptr' 0 seed
    pokeElemOff ptr' 1 gamma
```
(Type-checks, not tested)

Reply:
The equivalent to `MutGen` using `Storable` instead of `Prim` might look as follows:

```haskell
newtype IOGen g = IOGen (Ptr g)  

instance (RandomGen g, Storable g) =>
         MonadRandom (IOGen g) IO where   
  type Seed (IOGen g) = g  
  restore g = do
    ptr <- new g
    pure $ IOGen ptr
  save (IOGen ma) = peek ma
  uniformWord32R r = applyIOGen (genWord32R r)
  uniformWord64R r = applyIOGen (genWord64R r)
  uniformWord8 = applyIOGen genWord8
  uniformWord16 = applyIOGen genWord16
  uniformWord32 = applyIOGen genWord32
  uniformWord64 = applyIOGen genWord64
  uniformByteArray n = applyIOGen (genByteArray n)

applyIOGen :: Storable g => (g -> (a, g)) -> IOGen g -> IO a
applyIOGen f (IOGen ma) = do
  g <- peek ma
  case f g of
    (res, g') -> res <$ new g'
```

This is less general than `MutGen` since `Storable`/`Ptr` operations take place in `IO` rather than `PrimMonad`. The advantage is that `Storable` seems a lot easier to define than `Prim` for PRNGs.

Reply:
Using `new` is a guaranteed memory leak, even if you use `free` later on, so this code is certainly broken, `ForeignPtr` should be used instead. I do get what you are trying to say.

There are problems with `Storable` approach, when compared with `Prim`:

* `Prim` type doesn't force the memory allocated to be pinned, therefore small types with 128bits like pure generators will not contribute to memory fragmentation
* `Prim` is easier to use with `PrimMonad`, there is a bit of unsafe trickery needed to get `Storable` functions that work in `IO` to work in `PrimMonad`. Although it can be done, if we are gonna use `Storable`, I'd rather for it work in IO 

Benefit of using `Storable` are: 

* hands down easier to write instances. Here is an example for a 128bit `Prim` instance:
https://github.com/erikd/wide-word/blob/4a98ed263d6dacb92d317782db311599b79d7f1d/src/Data/WideWord/Word128.hs#L152-L172
* Ability to pass generator state to the C world.

The good thing is, we can add implementation for `Storable` as well, they aren't mutually exclusive. I can see both being useful with different generators in different scenarios.

Reply:
Oh, and `applyIOGen` is totally broken :)

It should be:
```haskell
    (res, g') -> liftIO $ res <$ poke ptr g'
```

Reply:
> Oh, and `applyIOGen` is totally broken :)
> 
> It should be:
> 
> ```haskell
>     (res, g') -> liftIO $ res <$ poke ptr g'
> ```

:man_facepalming: thanks - fixed.

Reply:
Mostly copied from https://github.com/erikd/wide-word/blob/4a98ed263d6dacb92d317782db311599b79d7f1d/src/Data/WideWord/Word128.hs#L152-L172:

```haskell
unInt :: Int -> Int#
unInt (I# i#) = i# 
            
instance Prim SMGen where
  sizeOf# _ = 2# *# sizeOf# (undefined :: Word64)
  alignment# _ = 2# *# alignment# (undefined :: Word64)
  indexByteArray# arr# i# =
    let i2# = 2# *# i#
        seed = indexByteArray# arr# i2#
        gamma = indexByteArray# arr# (i2# +# unInt 1)
    in SMGen seed gamma
  indexOffAddr# addr# i# =
    let i2# = 2# *# i#
        seed = indexOffAddr# addr# i2#
        gamma = indexOffAddr# addr# (i2# +# unInt 1)
    in SMGen seed gamma
  readByteArray# arr# i# =
    \s0 -> case readByteArray# arr# i2# s0 of             
      (# s1, seed #) -> case readByteArray# arr# (i2# +# unInt 1) s1 of
        (# s2, gamma #) -> (# s2, SMGen seed gamma #)
    where i2# = 2# *# i#     
  readOffAddr# addr# i# =
    \s0 -> case readOffAddr# addr# i2# s0 of
      (# s1, seed #) -> case readOffAddr# addr# (i2# +# unInt 1) s1 of
        (# s2, gamma #) -> (# s2, SMGen seed gamma #)
    where i2# = 2# *# i#
  writeByteArray# arr# i# (SMGen seed gamma) =
    \s0 -> case writeByteArray# arr# i2# seed s0 of             
      s1 -> case writeByteArray# arr# (i2# +# unInt 1) gamma s1 of
        s2 -> s2
    where i2# = 2# *# i#
  writeOffAddr# addr# i# (SMGen seed gamma) =
    \s0 -> case writeOffAddr# addr# i2# seed s0 of               
      s1 -> writeOffAddr# addr# (i2# +# unInt 1) gamma s1   
    where i2# = 2# *# i#
  setByteArray# = defaultSetByteArray#  
  setOffAddr# = defaultSetOffAddr#
```

(Type-checks, untested, ignores endianness)

So `Prim` instances are doable if cumbersome.

Reply:
Dammit, [I commented](https://github.com/idontgetoutmuch/random/pull/16#discussion_r388878177) on a incorrect PR. This exact thing is already implemened:

In case you wanna experiment with MutGen and `Prim` instance for SMGen here is a branch that has it: https://github.com/lehins/splitmix/tree/new-random

Also the benchmarks that I ran are here: https://github.com/lehins/splitmix/tree/new-random